### PR TITLE
[REVIEW] Port NVStrings attributes operations to cudf strings column

### PR DESCRIPTION
**Pull request by davidwendt**
This is a continuation of porting NVStrings operations to the new cudf column design.
This is PR is dependent on #2971 for some common utility functions.
It will only include the `characters_counts`, `bytes_counts`, and `code_points` functions.
Other string attributes functions will be in a follow-on PR.
This code was removed from the original #2811 PR and has been updated to for the current code base.

Progress on the NVStrings port overall is kept in the following document
https://nvidia-my.sharepoint.com/:x:/p/dwendt/EdaXvCXQ1r5Kq83KQuKWegIBEIjzK3FTptTFI4xsO2AFWg?e=ZtMXhJ

### CODE REVIEW

File: cpp/include/cudf/strings/attributes.hpp [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+ * @param output_type The integer type used for creating the output column.
+ *        Default type is INT32.
+ * @param mr Resource for allocating device memory.
+ * @param stream Stream to use for any kernels in this function.
+ * @return New column with lengths for each string.
+ */
+std::unique_ptr<cudf::column> characters_counts( strings_column_view strings,
+                                                 data_type output_type = data_type{INT32},
+                                                 rmm::mr::device_memory_resource* mr = rmm::mr::get_default_resource(),
+                                                 cudaStream_t stream = 0);
```
Comment (harrism) [inline cpp/include/cudf/strings/attributes.hpp]: streams before mrs.

Comment (davidwendt) [reply cpp/include/cudf/strings/attributes.hpp]: Found this in the transition guide: https://github.com/rapidsai/cudf/blob/branch-0.11/cpp/docs/TRANSITIONGUIDE.md

> The recommended pattern for doing this is to make the definition of the external API invoke an internal API in the detail namespace. The internal detail API will have all the same parameters, plus a cudaStream_t parameter at the end defaulted to 0. The implementation should be wholly contained in the detail API definition and use only asynchronous versions of CUDA APIs with the defaulted stream parameter.

Comment (davidwendt) [reply cpp/include/cudf/strings/attributes.hpp]: And so it looks like I need to add detail surrogate functions for all my APIs.

Comment (harrism) [reply cpp/include/cudf/strings/attributes.hpp]: Even so, in most of the APIs that have been ported so far, `stream` is before `mr` in the parameter list. Is there a reason to reverse it here?

Comment (davidwendt) [reply cpp/include/cudf/strings/attributes.hpp]: I removed the parameter and placed it in a detail function as per the transition guide.

File: cpp/include/cudf/strings/attributes.hpp [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+ * specified strings column. Each row value will be the number of
+ * characters in the corresponding string.
+ *
+ * Any null string will result in a null entry for that row in the output column.
+ *
+ * @param strings Strings instance for this operation.
+ * @param mr Resource for allocating device memory.
+ * @return New INT32 column with lengths for each string.
+ */
+std::unique_ptr<cudf::column> characters_counts( strings_column_view strings,
```
Comment (harrism) [inline cpp/include/cudf/strings/attributes.hpp]: I think the item counted should be singular. There are multiple counts returned, which is already indicated by the plural counts.

```suggestion
std::unique_ptr<cudf::column> character_counts( strings_column_view strings,
```

File: cpp/include/cudf/strings/attributes.hpp [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+ * specified strings column. Each row value will be the number of
+ * bytes in the corresponding string.
+ *
+ * Any null string will result in a null entry for that row in the output column.
+ *
+ * @param strings Strings instance for this operation.
+ * @param mr Resource for allocating device memory.
+ * @return New INT32 column with the number of bytes for each string.
+ */
+std::unique_ptr<cudf::column> bytes_counts( strings_column_view strings,
```
Comment (harrism) [inline cpp/include/cudf/strings/attributes.hpp]:
```suggestion
std::unique_ptr<cudf::column> byte_counts( strings_column_view strings,
```

**Comment by harrism** [comment]
I suggested `character_counts` and `byte_counts`, but you changed it to `characters_count` and `bytes_count`. Do they produce a single count, or one per row? One per row => "counts".

rerun tests.

**Comment by davidwendt** [comment]
> I suggested `character_counts` and `byte_counts`, but you changed it to `characters_count` and `bytes_count`. Do they produce a single count, or one per row? One per row => "counts".
> 
Sorry, misread your comment. It is a single count. One count per row. Each row is a string. The count is either the count of bytes or the count of characters depending on the API.

**Comment by harrism** [comment]
Ugh. I guess the question should be whether APIs are verb names (after what action they perform) or noun names (after what they produce).

If verb: `count_characters` : it counts characters for all rows.
If noun: `character_counts` : it returns the character counts for all rows.

I don't see how `characters_count` fits in, since it returns multiple counts.

I think I like `count_characters` and `count_bytes` best since they are lexicographically closer.

**Pull request merged**